File: webmvp/src/lib/keyUtils.ts
import { ALL_KEYS, type Key } from "@/lib/engine";

export function keyIndex(key: Key): number {
  return ALL_KEYS.indexOf(key);
}

export function transposeKeyBy(key: Key, semitones: number): Key {
  const idx = (keyIndex(key) + semitones + ALL_KEYS.length) % ALL_KEYS.length;
  return ALL_KEYS[idx];
}

export function isKey(value: string): value is Key {
  return (ALL_KEYS as readonly string[]).includes(value);
}

File: mobile/lib/domain/key_utils.dart
import 'package:lf_chords/domain/engine.dart';

export 'engine.dart' show allKeys, allKeysSet, isValidKey;

int keyIndex(String key) => allKeys.indexOf(key);

String transposeKeyBy(String key, int semitones) {
  final idx = (keyIndex(key) + semitones + allKeys.length) % allKeys.length;
  return allKeys[idx];
}

bool isKey(String value) => isValidKey(value);
